### DISCUSSION TITLE
Processing Batched Transactions w/ Some Fails

### DISCUSSION THREAD

**Discussion by setzeus** [stacks-sbtc/sbtc#261] [Design]
Surfaced in #243, there's a discrepancy in how we handle submitting batched transactions (deposits | withdrawals) on-chain. 

Currently, in both user paths, a signer submits a list of either deposits | withdrawals & *all* individual transactions should pass. If a single one fails, then they all fail with an error message pointing to the failed entry in the list. 

As mentioned by @djordon, this might not necessarily be the best way to handle this logic. Is it a better alternative to complete the transactions even if one or more of them fail?

**Comment by djordon** [comment]
We spoke about this briefly offline. After some thought here is what I think:
1. Whatever we do, we should adopt the same behavior regardless of whether we use the "list version" of the contract functions or the "single version" of the contract functions (say, called in a loop).
2. We should not halt all sBTC operations if there is a contract call failure[^1].
3. @hstove brought up a good point about how transaction failures roll back state updates. We want this for the failures, so if we do not short-circuit then we need to be careful to ensure that things are in the right state on successes and failures.

With the short-circuiting approach, we'd have to retry the ones that are known to have succeeded, as well as the ones that are unknown (this follows from (1) and (2)). I think this means that we should return enough information in the error response for us to figure out which ones failed. For v1, I think that means the signer should just use the "single version" of the contract call. The "single version" approach makes things super simple on the signer side. It's more costly because of fees, so maybe we can do the ideal thing if we have time.


[^1]: So if a contract call for a withdrawal request fails for someone, then we should continue to try sBTC operations for other individuals. Maybe there is a set of withdrawal contract call failures where we want to stop all other withdrawals. But I'm having trouble imagining what that is. Same thing for deposit requests. In any case, I think this might deserve its own discussion.
